### PR TITLE
ProjectileInfo

### PR DESCRIPTION
**Pull request by Sergeanur**

### CODE REVIEW

File: src/weapons/ProjectileInfo.cpp [hunk 1]
```diff
@@ -1,17 +1,302 @@
...
+		float vy = 1.25f;
+		time = CTimer::GetTimeInMilliseconds() + 1400;
+		if (ped->IsPlayer()) {
+			matrix.GetForward() = TheCamera.Cams[TheCamera.ActiveCam].Front;
+			matrix.GetUp() = TheCamera.Cams[TheCamera.ActiveCam].Up;
+			matrix.GetRight() = CrossProduct(TheCamera.Cams[TheCamera.ActiveCam].Up, TheCamera.Cams[TheCamera.ActiveCam].Front);
+			matrix.GetPosition() = pos;
+		} else if (ped->m_pSeekTarget != nil) {
+				float ry = CGeneral::GetRadianAngleBetweenPoints(1.0f, ped->m_pSeekTarget->GetPosition().z, 1.0f, pos.z);
+				float rz = atan2(-ped->GetForward().x, ped->GetForward().y);
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: `Atan2`

File: src/weapons/ProjectileInfo.cpp [hunk 2]
```diff
@@ -1,17 +1,302 @@
...
+	{
+	case WEAPONTYPE_ROCKETLAUNCHER:
+	{
+		float vy = 1.25f;
+		time = CTimer::GetTimeInMilliseconds() + 1400;
+		if (ped->IsPlayer()) {
+			matrix.GetForward() = TheCamera.Cams[TheCamera.ActiveCam].Front;
+			matrix.GetUp() = TheCamera.Cams[TheCamera.ActiveCam].Up;
+			matrix.GetRight() = CrossProduct(TheCamera.Cams[TheCamera.ActiveCam].Up, TheCamera.Cams[TheCamera.ActiveCam].Front);
+			matrix.GetPosition() = pos;
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: nit: here I thought about adding setter `SetPosition` and using later?

File: src/weapons/ProjectileInfo.cpp [hunk 3]
```diff
@@ -1,17 +1,302 @@
...
+	float elasticity = 0.75f;
+	CPed* ped = (CPed*)entity;
+	int time;
+	CVector velocity;
+
+	switch (weapon)
+	{
+	case WEAPONTYPE_ROCKETLAUNCHER:
+	{
+		float vy = 1.25f;
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: maybe variables named `vX`, ,`vY`, `vZ`? @aap

File: src/weapons/ProjectileInfo.cpp [hunk 4]
```diff
@@ -1,17 +1,302 @@
...
+	case WEAPONTYPE_FLAMETHROWER:
+		Error("Undefined projectile type, AddProjectile, ProjectileInfo.cpp");
+		break;
+	case WEAPONTYPE_MOLOTOV:
+	{
+		time = CTimer::GetTimeInMilliseconds() + 2000;
+		float scale = 0.22f * speed + 0.15f;
+		if (scale >= 0.2f)
+			scale = 0.2f;
+		float angle = atan2(-ped->GetForward().x, ped->GetForward().y);
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: `Atan2`

File: src/weapons/ProjectileInfo.cpp [hunk 5]
```diff
@@ -1,17 +1,302 @@
...
+		if (speed != 0.0f)
+			scale = 0.22f * speed + 0.15f;
+		float angle = atan2(-ped->GetForward().x, ped->GetForward().y);;
+		matrix.SetTranslate(0.0f, 0.0f, 0.0f);
+		matrix.RotateZ(angle);
+		matrix.GetPosition() += pos;
+		SpecialCollisionResponseCase = COLLRESPONSE_UNKNOWN5;
+		velocity.x = -1.0f * scale * Sin(angle);
+		velocity.y = scale * Cos(angle);
+		velocity.z = (0.4f * speed + 0.4f) * scale;
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: unoptimized operator?

File: src/weapons/ProjectileInfo.cpp [hunk 6]
```diff
@@ -1,17 +1,302 @@
...
+		velocity.z = (0.2f * speed + 0.4f) * scale;
+		break;
+	}
+	case WEAPONTYPE_GRENADE:
+	{
+		time = CTimer::GetTimeInMilliseconds() + 2000;
+		float scale = 0.0f;
+		if (speed != 0.0f)
+			scale = 0.22f * speed + 0.15f;
+		float angle = atan2(-ped->GetForward().x, ped->GetForward().y);;
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: etc

File: src/weapons/ProjectileInfo.cpp [hunk 7]
```diff
@@ -1,17 +1,302 @@
...
+	for (int i = 0; i < ARRAY_SIZE(ms_apProjectile); i++) {
+		if (gaProjectileInfo[i].m_bInUse) {
+			if (gaProjectileInfo[i].m_eWeaponType == WEAPONTYPE_ROCKETLAUNCHER || gaProjectileInfo[i].m_eWeaponType == WEAPONTYPE_MOLOTOV || gaProjectileInfo[i].m_eWeaponType == WEAPONTYPE_GRENADE) {
+				CVector &pos = ms_apProjectile[i]->GetPosition();
+				if (pos.x >= x1 && pos.x <= x2 && pos.y >= y1 && pos.y <= y2 && pos.z >= z1 && pos.z <= z2) {
+					result = true;
+					if (remove) {
+						gaProjectileInfo[i].m_bInUse = false;
+						CWorld::Remove(ms_apProjectile[i]);
+						delete ms_apProjectile[i];
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: it's more like fix eventual bugs, but it could be set as `nil`.

File: src/weapons/ProjectileInfo.cpp [hunk 8]
```diff
@@ -1,17 +1,302 @@
...
+}
+
+void
+CProjectileInfo::RemoveAllProjectiles()
+{
+	for (int i = 0; i < ARRAY_SIZE(ms_apProjectile); i++) {
+		if (gaProjectileInfo[i].m_bInUse) {
+			gaProjectileInfo[i].m_bInUse = false;
+			CWorld::Remove(ms_apProjectile[i]);
+			delete ms_apProjectile[i];
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.cpp]: also

File: src/weapons/ProjectileInfo.h [hunk 9]
```diff
@@ -10,20 +10,23 @@ class CProjectileInfo
...
+	static void RemoveNotAdd(CEntity *entity, eWeaponType weaponType, CVector pos);
 	static bool RemoveIfThisIsAProjectile(CObject *pObject);
 	static void RemoveAllProjectiles(void);
+	static void Update();
 	static bool IsProjectileInRange(float x1, float x2, float y1, float y2, float z1, float z2, bool remove);
 };
 
-extern CProjectileInfo* gaProjectileInfo;
\ No newline at end of file
+extern CProjectileInfo (&gaProjectileInfo)[32];
```
Comment (ShFil119) [inline src/weapons/ProjectileInfo.h]: 32 defined as `const`/`DEFINE`, it's also used in cpp

File: src/math/Vector.h [hunk 10]
```diff
@@ -109,7 +109,7 @@ DotProduct(const CVector &v1, const CVector &v2)
 	return v1.x*v2.x + v1.y*v2.y + v1.z*v2.z;
 }
 
-inline CVector
+inline const CVector
```
Comment (ShFil119) [inline src/math/Vector.h]: not sure if needed

Comment (Sergeanur) [reply src/math/Vector.h]: Restores original behavior

**Pull request merged**